### PR TITLE
Lite: Fully_connected op bug fixed and code refactored

### PR DESCRIPTION
**Pull request by ANSHUMAN87**
1:> TF_LITE_MACRO_DISPATCH --> remove this macro unused
2:> Unnecessary check removed as it is clearly evident from code above
3:> Shift negation is not required explicitly when computed, it should be binded with code where it is used, if needed
4:> Confusing comment removed
5:> Error message improvised
6:> Code refactored to accommodate multiple if() else ()
7:> Move data type check to one central place, to improve code readability
8:> 2 Bug fixed as part of above change
       

-  The code was prone to significant data loss, if user has given Input as int16 type which was explicitly type casted to int8/uint8.
-  Null pointer access when Bias is not present at EvalShuffledQuantized()

9:> If condition modified to imply quantized input

### CODE REVIEW

**Comment by jdduke** [comment]
Please avoid coupling refactoring changes with actual behavior changes and/or bug fixes. Can you split this CL into different pieces accordingly (with the highest priority being the bug fix)? Thanks!

File: tensorflow/lite/kernels/fully_connected.cc [hunk 1]
```diff
@@ -64,6 +64,48 @@ constexpr int kBiasTensor = 2;
 constexpr int kOutputTensor = 0;
 constexpr int kShuffledInputWorkspaceTensor = 1;
 
+inline bool CheckDtype(const TfLiteTensor* input, const TfLiteTensor* filter,
+                       const TfLiteTensor* bias, TfLiteTensor* output,
+                       TfLiteFullyConnectedParams* params) {
+  bool is_Quantized =
+      ((filter->type == kTfLiteUInt8) || (filter->type == kTfLiteInt8));
+  bool is_Hybrid = is_Quantized && (input->type == kTfLiteFloat32);
+  bool is_Shuffled =
```
Comment (shahzadlone) [inline tensorflow/lite/kernels/fully_connected.cc]: ```is_shuffled```  can be a const.

File: tensorflow/lite/kernels/fully_connected.cc [hunk 2]
```diff
@@ -64,6 +64,48 @@ constexpr int kBiasTensor = 2;
 constexpr int kOutputTensor = 0;
 constexpr int kShuffledInputWorkspaceTensor = 1;
 
+inline bool CheckDtype(const TfLiteTensor* input, const TfLiteTensor* filter,
+                       const TfLiteTensor* bias, TfLiteTensor* output,
+                       TfLiteFullyConnectedParams* params) {
+  bool is_Quantized =
+      ((filter->type == kTfLiteUInt8) || (filter->type == kTfLiteInt8));
+  bool is_Hybrid = is_Quantized && (input->type == kTfLiteFloat32);
```
Comment (shahzadlone) [inline tensorflow/lite/kernels/fully_connected.cc]: ```is_Hybrid``` can be a const.

File: tensorflow/lite/kernels/fully_connected.cc [hunk 3]
```diff
@@ -64,6 +64,48 @@ constexpr int kBiasTensor = 2;
...
+  bool is_Shuffled =
+      is_Quantized && (params->weights_format ==
+                       kTfLiteFullyConnectedWeightsFormatShuffled4x16Int8);
+  bool is_BiasNotFloatType = bias && (bias->type != kTfLiteFloat32);
+  bool is_BiasNotIntType = bias && (bias->type != kTfLiteInt32);
```
Comment (shahzadlone) [inline tensorflow/lite/kernels/fully_connected.cc]: same as above

File: tensorflow/lite/kernels/fully_connected.cc [hunk 4]
```diff
@@ -64,6 +64,48 @@ constexpr int kBiasTensor = 2;
...
+inline bool CheckDtype(const TfLiteTensor* input, const TfLiteTensor* filter,
+                       const TfLiteTensor* bias, TfLiteTensor* output,
+                       TfLiteFullyConnectedParams* params) {
+  bool is_Quantized =
+      ((filter->type == kTfLiteUInt8) || (filter->type == kTfLiteInt8));
+  bool is_Hybrid = is_Quantized && (input->type == kTfLiteFloat32);
+  bool is_Shuffled =
+      is_Quantized && (params->weights_format ==
+                       kTfLiteFullyConnectedWeightsFormatShuffled4x16Int8);
+  bool is_BiasNotFloatType = bias && (bias->type != kTfLiteFloat32);
```
Comment (shahzadlone) [inline tensorflow/lite/kernels/fully_connected.cc]: same as above

**Comment by ANSHUMAN87** [comment]
> Please avoid coupling refactoring changes with actual behavior changes and/or bug fixes. Can you split this CL into different pieces accordingly (with the highest priority being the bug fix)? Thanks!

@jdduke : Thanks for your suggestion, i will keep in mind for my future PRs. Because the PR contains multiple changes, i have consolidated the list of changes above with descriptions, if that helps. Otherwise should i close the current PR and raise multiple PRs, or raise multiple commits in one PR? Please let me know your convenience, TIA!

**Comment by jdduke** [comment]
>  Otherwise should i close the current PR and raise multiple PRs, or raise multiple commits in one PR? Please let me know your convenience, TIA!

Prefer multiple PRs. In practice, once this change gets approved, it is landed internally as an atomic commit. Better to have it split across multiple (real) commits than a single one.

**Comment by jdduke** [comment]
Also, it makes it easier to review and land if it focuses on a single issue/fix.

**Comment by ANSHUMAN87** [comment]
> Also, it makes it easier to review and land if it focuses on a single issue/fix.

@jdduke : I am closing this PR, to raise again with multiple PRs as suggested by you.
@shahzadlone : Thanks for your comments, i will consider those in my next PRs.

**Comment by ANSHUMAN87** [comment]
Closing to raise again.

**Pull request closed without merging**